File: Sockets_Programming_fall16/CS3251_PA1_MingjunXie/Sensor_tcp.java
import java.net.Socket;
import java.net.SocketException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Random;
public class Sensor_tcp {

  public static void main(String[] args) throws IOException {
    boolean debug = false;

    if (args.length != 10 || args.length != 11) {// Test for correct # of args
      throw new IllegalArgumentException("sensor-tcp usage: -s host name -p port number -u username -c password -r sensor value -d for debug mode");
      // throw an exception if args are not in the format
    }

    String server = args[1];
    int portNumber = Integer.parseInt(args[3]);
    String username = args[5];
    String password = args[7];
    String sensorVal = args[9];

    if (args[10] == "-d") {
        debug = true;
    }

    // if debug mode, the information will be printed out
    if (debug) {
        System.out.println("server host name:" + server);
        System.out.println("server port number:" + portNumber);
        System.out.println("username:" + username);
        System.out.println("password:" + password);        
        System.out.println("sensor value:" + sensorVal);
    }

    // Create socket that is connected to server on specified port
        Socket socket = new Socket(server, portNumber);
        System.out.println("Connected to server...sending string");        

    //need MD5 algorithm
    String challenge = "ABC" + new Random().nextInt(9999)*31%11 + 7;
    try {
        byte[] hash = new MD5().MD5ALG(username, password, challenge);
        InputStream in = socket.getInputStream();
        OutputStream out = socket.getOutputStream();
        out.write(hash);  // Send the encoded string to the server
    } catch (Exception e) {
        System.out.println("There's an error in the algorithm");
    }
 

    // Client sends "authentication request" message.
    //Server responds with a one time use, challenge value in the form of a random 64 character string. (You get to decide how this random string is generated.)
    //Client computes a MD5 hash of the string formed by concatenating the username, password and the random string sent by the server. Hash = MD5("username","password","challenge")
    //Client sends the clear text "username" and the resulting "Hash" to the server.
    //The server takes the username, finds the corresponding password on file, and performs the same MD5 calculation. It then compares the calculated Hash to the one sent by the client. 
    //If they match, the user has successfully authenticated. If no match, then authentication fails.



    // // Receive the same string back from the server
    // int totalBytesRcvd = 0;  // Total bytes received so far
    // int bytesRcvd;           // Bytes received in last read
    // while (totalBytesRcvd < data.length) {
    //   if ((bytesRcvd = in.read(data, totalBytesRcvd,  
    //                     data.length - totalBytesRcvd)) == -1)
    //     throw new SocketException("Connection closed prematurely");
    //   totalBytesRcvd += bytesRcvd;
    // }  // data array is full

    // System.out.println("Received: " + new String(data));

    socket.close();  // Close the socket and its streams
  }
}

File: Sockets_Programming_fall16/CS3251_PA1_MingjunXie/sensor-tcp.py
#!/usr/bin/python
'''
Created on 09/30/2016

@author: Mingjun Xie
'''
import socket   # for sockets
import sys  # for exit
import getopt   # for handling command line arguments
import Util


def runTCPClient():  
    # Create an INET, STREAMing socket (TCP)
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    except socket.error:
        print 'Failed to create socket'
        sys.exit()
    if (d):
        print 'Socket Created'

    # Get the server IP
    try:
        serverIP = socket.gethostbyname(server)
        if (d):
            print "serverIP: ", serverIP
        # Handles passing in an IP address as well as a DNS name
    except socket.gaierror:
        # Host name could not resolve
        print 'Hostname could not be resolved. Exiting'
        s.close()
        sys.exit()
     
    # Connect to the server
    try:
        s.connect((serverIP , port))
        if (d):
            print 'Socket Connected to ' + server + ' on ip ' + serverIP
    except socket.error, msg:
        print 'Connection error: ', msg
        s.close()
        sys.exit()

    # Send the request to remote server

    try :
        requestMsg = "request"
        # Send the request string
        s.send(requestMsg)
    except socket.error:
        # Send failed
        print 'Send failed'
        s.close()
        sys.exit()

    # Receive the challenge string from the server
    challenge = ""
    try:
        challenge = s.recv(4096)
    except socket.error, msg:
        print "Challenge string accept error:" , msg
        s.close()
        sys.exit()

    hashMD5 = Util.MD5(username, password, challenge)
    try :
        s.sendall(username + "," + hashMD5 + "," + sensorVal)
    except socket.error, msg:
        print "Authentication information send failed: ", msg
        s.close()
        sys.exit()

    response = s.recv(4096)
    print "Reply from server: \n", response
    # Close the socket, signals that the application is ready to close the TCP connection
    print 'Socket Closed to ' + server + ' on ip ' + socket.gethostbyname(socket.gethostname())
    s.close()


if __name__ == '__main__':
    try:
        opts, args = getopt.getopt(sys.argv[1:],"s:p:u:c:r:d",["server=","port=","username=","password=","sensorvalue="])
    except getopt.GetoptError as err:
        print err
        print 'Usage:Python Sensor.tcp.py -s host name -p port number -u username -c password -r sensor value -d debug mode'

    d = False
    server = None
    port = None
    username = None
    password = None
    sensorVal = None

    for opt, arg in opts:
        if opt == '-h':
            print 'usage:'
            print sys.argv[0] + ' -s <host name> -p <port number> -u <username> -c <password> -r <sensorvalue> -d <debug mode>'
            sys.exit()
        elif opt in ("-s", "--server"):
            server = arg
        elif opt in ("-p", "--port"):
            if arg.isdigit():
                port = int(arg)
            else:
                print "Port number is not in digits."
                sys.exit()
        elif opt in ("-u", "--username"):
            username = arg
        elif opt in ("-c", "--password"):
            password = arg
        elif opt in ("-r", "--sensorvalue"):
            sensorVal = arg
        elif opt in ("-d"):
            d = True;

    if (server == None) or (port == None) or (username == None) or (password == None) or (sensorVal == None) :
        print "Please don't let server, port number, username, password or sensorvalue be null"
        sys.exit()

    if (port not in range(1024,65536)):
        print "Bad port number: port number should goes between 1024 and 65536."
        sys.exit()
    if d:
        print "server host name: ", server
        print "server port number: ", port
        print "username:", username
        print "password:", password     
        print "sensor value:", sensorVal
    runTCPClient()
